app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb

```diff
@@ -124,6 +135,43 @@ def get_images
         process_collection(images, :vms) { |image| parse_image(image) }
       end
 
+      def get_security_groups
+        security_groups = gather_data_for_this_region(@nsg)
+
+        process_collection(security_groups, :security_groups) do |sg|
+          parse_security_group(sg)
+        end
+
+        get_firewall_rules
+      end
+
+      # To get the list of security groups for a VM we need to find any
+      # NIC's associated with the security group, and then get the VM's
+      # associated with the NIC's.
+      #
+      def get_security_groups_for_vm(vm)
+        group_array = []
+
+        security_groups.each do |sg|
+          sg.properties.network_interfaces.each do |nic|
+            resource_group = nic.id[%r{resourceGroups/(.*?)/}, 1]
+            nic = @nis.get(File.basename(nic.id), resource_group)
+            if nic.properties['virtualMachine'] and nic.properties.virtual_machine.id == vm.id
+              group_array << sg
+            end
+          end
+        end
+
+        group_array
```

Comment:
@djberg96 and I chatted about this offline ... 

I looked this over again, and I'm uncomfortable with this approach ... even with the notion of coming back around to clean it up :/

Here's how I think this could be cleaned up without *too* much refactoring:

1. introduce `get_vm_nics` that just gathers the `nic_profiles` currently [gathered by `populate_hardware_has_with_networks`](https://github.com/djberg96/manageiq/blob/05372005b3de5d7cba13c71c8425edd081f2cc8c/app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb#L368)
2. pass the nics from step 1 into `populate_hardware_has_with_networks`
3. pass the same nics into `get_security_groups_for_vm` (which should be renamed to `get_security_groups_for_instance` to stay consistent with calling these `"instances"`)

Then, it's a matter of comparing the `security_group`s' `nic`s' IDs to the `instance`s' `nic`s' IDs.  Then, gathering the data for the nics is done in one place and not in multiple places (multiple times, even).

The only question I see left open from that is in [ `populate_hardware_has_with_networks`](https://github.com/djberg96/manageiq/blob/05372005b3de5d7cba13c71c8425edd081f2cc8c/app/models/manageiq/providers/azure/cloud_manager/refresh_parser.rb#L368).  That method currently creates a new `NetworkInterfaceService` for each nic associated with a VM.  Can each nic actually belong to a different subscription?  If not, that seems like overkill and could probably be done away with.